src/transformers/models/bloom/modeling_flax_bloom.py

```diff
@@ -0,0 +1,771 @@
+        attention_mask=None,
+        deterministic: bool = True,
+        init_cache: bool = False,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+    ):
+        all_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        if self.use_scan:
+            # since all decoder layers are the same, we use nn.scan directly
+            # assert not output_attentions, "cannot use `scan` with `output_attentions` set to `True`"
+            # assert not output_hidden_states, "cannot use `scan` with `output_hidden_states` set to `True`"
+            hidden_states = (hidden_states,)
+
+            hidden_states, _ = scan_with_axes(
+                FlaxBloomBlock,
+                variable_axes={"params": 0, "cache": 0},
+                split_rngs={"params": True, "dropout": True},
+                in_axes=(nn.broadcast, nn.broadcast, 0, nn.broadcast, nn.broadcast, nn.broadcast),
+                length=self.config.num_hidden_layers,
+            )(self.config, dtype=self.dtype, use_scan=True, name="FlaxBloomBlockLayers")(
+                hidden_states,
+                alibi,
+                attention_mask,  # kwargs not supported by scan
+                jnp.arange(self.config.num_hidden_layers),
+                None,
+                deterministic,
+                init_cache,
+            )
+            hidden_states = hidden_states[0]
+
+        else:
+            for layer_number in range(self.config.num_hidden_layers):
+                if output_hidden_states:
+                    all_hidden_states += (hidden_states,)
+
+                layer_outputs = FlaxBloomBlock(self.config, name=str(layer_number), dtype=self.dtype, use_scan=False)(
+                    hidden_states,
+                    alibi=alibi,
+                    attention_mask=attention_mask,
+                    deterministic=deterministic,
+                    init_cache=init_cache,
+                    output_attentions=output_attentions,
+                    layer_number=layer_number,
+                )
+                hidden_states = layer_outputs[0]
+
+                if output_attentions:
+                    all_attentions += (layer_outputs[1],)
+
+        # this contains possible `None` values - `FlaxBloomModule` will filter them out
+        outputs = (hidden_states, all_hidden_states, all_attentions)
+
+        return outputs
+
+
+class FlaxBloomModule(nn.Module):
+    config: BloomConfig
+    dtype: jnp.dtype = jnp.float32
+    use_scan: bool = False
+
+    def setup(self):
+        self.embed_dim = self.config.hidden_size
+
+        embedding_init = jax.nn.initializers.normal(stddev=self.config.initializer_range)
+
+        # word embeddings (no positional embedding layer)
+        self.word_embeddings = nn.Embed(
+            self.config.vocab_size,
+            self.embed_dim,
+            embedding_init=embedding_init,
+            dtype=self.dtype,
```

Comment:
We shouldn't pass the `dtype` to embeds, this will keep the embed weights in that `dtype`,  but it's fine if that's what we want here.

Reply:
I think that this is what we want here 💪

Reply:
We should pass the `dtype` arg to keep the weights in bfloat16 for half precision inference.

Also related: https://github.com/huggingface/transformers/pull/18462